Extends scope of transactions to rollback when using dry_run=True

kreatemore opened the pull request:
**Problem**

I believe the changes below address and fix the following:
https://github.com/django-import-export/django-import-export/issues/1099
https://github.com/django-import-export/django-import-export/issues/1447
https://github.com/django-import-export/django-import-export/issues/1078

and our similar problem we're facing internally (at work) which is exactly the same as https://github.com/django-import-export/django-import-export/issues/1078#issuecomment-600225790

**Solution**

Instead of creating a `savepoint` before inserting rows and then rolling back to said `savepoint` (if applicable), I'm proposing to implicitly mark the outermost transaction to be rolled back if `dry_run=True and use_transactions=True`. This should mark related functions awaiting execution `on_commit` to be considered "rolled back". 

I'm not a 100% sure I fully understand why this happens, but during my investigation of this issue I've found that the manually created savepoint didn't exist in `connection.savepoint_ids`, but rolling back to it was successful, while another savepoint was added to `connection.savepoint_ids` (these were the auxiliary actions, like signal handlers, etc), but there was no related savepoint in the database. I suspect it might be somewhat due to threading, but I'm not experienced enough to be able to tell confidently if that's the case.

**Important!** While I think this is more aligned with what end-users would expect to happen when rolling back, and existing tests still pass, this might be considered as a breaking change. Should this be based to the v3-release branch instead?

**Acceptance Criteria**

Have you written tests? ✅  
Have you included screenshots of your changes if applicable? N/A
Did you document your changes? I sure hope so!

Thanks for the review in advance!

matthewhegarty commented:
@pokken-magic would appreciate your input as you raised #1447

matthewhegarty commented:
Thanks for submitting this PR - I'll review in detail when I get a chance.

> Should this be based to the v3-release branch instead?

Yes I think that if this goes in, it would be safer to go into v3

pokken-magic commented:
Happy to take a look at this one when it's based to 3 and passing tests. Looks very very promising though! Thanks for doing it.

kreatemore commented:
- [x] Fixed imports formatting
- [x] PR now points to `release-3-x`
- [x] Updated my fork's `main` to be `release-3-x` so we don't have to worry about merge conflicts in this PR

I think the PR is ready for another round of CI, fingers crossed all tests pass 🤞🏼

matthewhegarty commented:
I think we'll need to take a step back and try to understand exactly what's going on before we can merge.

From what I have read from previous issues, the purpose of this fix is to stop the `post_save()` signal from being called twice.  Is that correct?

If you have any more insight into how this works, or how exactly it fixes the issue I would be interested to understand in more detail.

For reference, this is the implementation of `django.db.transaction.set_rollback()`

```python
def set_rollback(rollback, using=None):
    """
    Set or unset the "needs rollback" flag -- for *advanced use* only.

    When `rollback` is `True`, trigger a rollback when exiting the innermost
    enclosing atomic block that has `savepoint=True` (that's the default). Use
    this to force a rollback without raising an exception.

    When `rollback` is `False`, prevent such a rollback. Use this only after
    rolling back to a known-good state! Otherwise, you break the atomic block
    and data corruption may occur.
    """
    return get_connection(using).set_rollback(rollback)
```

matthewhegarty commented on tests/core/tests/test_resources.py:
```diff
@@ -1354,6 +1355,39 @@ def test_rollback_on_validation_errors_true(self):
         # Ensure the rollback has worked properly, no instances were created.
         self.assertFalse(Author.objects.exists())
 
+    def test_rollback_auxiliary_on_commit_transactions_when_dry_run(self):
```
We probably need an additional test to ensure that rollback is not called when `using_transactions` is False

kreatemore commented:
> From what I have read from previous issues, the purpose of this fix is to stop the post_save() signal from being called twice. Is that correct?

I think that's correct

>If you have any more insight into how this works, or how exactly it fixes the issue I would be interested to understand in more detail.

What I think is happening is that the savepoints in that the current implementation limits the scope of the rollback to the inserts made during the import preview: we check if the to-be-imported rows could be saved to the DB to ensure there are no integration errors etc, then we only rollback **those** inserts. In the meanwhile signals have already been dispatched for related (background) tasks, but they are exempt from the rollbacks, and executed as if we had really saved the rows and not just checked if we could.

matthewhegarty commented:
I added the following to `tests/core/admin.py`:

```python
@receiver(post_save, sender=Author)
def post_save_callback(sender, **kwargs):
    print("Save called")
```

I then imported the attached Author via the Example App.  With this fix we would expect the call back to be called once, in fact it gets called twice.  It looks like the change prevents the new row from being written to the db before the "confirm" step, but the signal gets sent twice.  Can you reproduce?  Have I missed something here?

[author.csv](https://github.com/django-import-export/django-import-export/files/9338398/author.csv)

matthewhegarty commented:
AFAIK, [setting a flag on the instance](https://stackoverflow.com/a/71625152/39296) would be the only way to prevent the `post_save` signal being called twice.

pokken-magic commented:
Fwiw I am not sure this solves 1447 because it's a peculiarity of file fields not directly connected to the save handlers (I don't think anyway). But it seems like a good idea if it can be made to work. Nothing else to add beyond Matthew's findings.  

Because the issue w 1447 is ultimately due to serializing files, which our library doesn't support, it's possible I should just close that one tbh.

kreatemore commented:
@matthewhegarty I'm starting to think maybe this "fix" was too specific for our use-case as we usually expect signals to be called on `transaction.on_commit`

For example:
```diff
@receiver(post_save, sender=Author)
def post_save_callback(sender, **kwargs):
-    print("Save called")
+    transaction.on_commit(lambda: print("Save called"))
```

That's very specific and I guess that'd be too much to impose on the users of the library.

I'll try to spend some time on this (thanks for pointing me to author.csv + the test django instance), but unfortunately covid is kicking my butt atm, so unsure when will I be able to pick this up again. In the meantime feel free to close this PR and I'll come back with another one if I can find a better solution
